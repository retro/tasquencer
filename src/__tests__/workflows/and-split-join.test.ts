import { Effect } from 'effect';
import { it } from 'vitest';

import { Builder, IdGenerator, Service } from '../../index.js';
import { getEnabledTaskNames, makeIdGenerator } from '../shared.js';

it('runs a net with "and" split and "and" join', ({ expect }) => {
  const workflowDefinition = Builder.workflow()
    .withName('checkout')
    .startCondition('start')
    .task('scan_goods', Builder.emptyTask())
    .task('pay', Builder.emptyTask().withSplitType('and'))
    .task('pack_goods', Builder.emptyTask())
    .task('issue_receipt', Builder.emptyTask())
    .task('check_goods', Builder.emptyTask().withJoinType('and'))
    .endCondition('end')
    .connectCondition('start', (to) => to.task('scan_goods'))
    .connectTask('scan_goods', (to) => to.task('pay'))
    .connectTask('pay', (to) => to.task('pack_goods').task('issue_receipt'))
    .connectTask('pack_goods', (to) => to.task('check_goods'))
    .connectTask('issue_receipt', (to) => to.task('check_goods'))
    .connectTask('check_goods', (to) => to.condition('end'));

  const program = Effect.gen(function* () {
    const idGenerator = makeIdGenerator();

    const service = yield* workflowDefinition.build().pipe(
      Effect.flatMap((workflow) => Service.initialize(workflow)),
      Effect.provideService(IdGenerator, idGenerator)
    );

    expect(yield* service.getState()).toMatchSnapshot();

    yield* service.startRootWorkflow();
    const state1 = yield* service.getState();
    expect(state1).toMatchSnapshot();
    expect(getEnabledTaskNames(state1)).toEqual(new Set(['scan_goods']));

    yield* service.startTask('scan_goods', {});
    const state2 = yield* service.getState();
    expect(state2).toMatchSnapshot();
    expect(getEnabledTaskNames(state2)).toEqual(new Set(['pay']));

    yield* service.startTask('pay', {});
    const state3 = yield* service.getState();
    expect(state3).toMatchSnapshot();
    expect(getEnabledTaskNames(state3)).toEqual(
      new Set(['pack_goods', 'issue_receipt'])
    );

    yield* service.startTask('pack_goods', {});
    const state4 = yield* service.getState();
    expect(state4).toMatchSnapshot();
    expect(getEnabledTaskNames(state4)).toEqual(new Set(['issue_receipt']));

    yield* service.startTask('issue_receipt', {});
    const state5 = yield* service.getState();
    expect(state5).toMatchSnapshot();
    expect(getEnabledTaskNames(state5)).toEqual(new Set(['check_goods']));

    yield* service.startTask('check_goods', {});
    const state6 = yield* service.getState();
    expect(state6).toMatchSnapshot();
    expect(state6.workflows[0]?.state).toBe('completed');
  });

  Effect.runSync(program);
});
