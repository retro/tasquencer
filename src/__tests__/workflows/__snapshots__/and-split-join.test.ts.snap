// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runs a net with "and" split and "and" join 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:scan_goods->pay",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pay->pack_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pay->issue_receipt",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pack_goods->check_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:issue_receipt->check_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
  ],
  "tasks": [
    {
      "generation": 0,
      "name": "scan_goods",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "pay",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "pack_goods",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "issue_receipt",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "check_goods",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "checkout",
      "parent": null,
      "state": "initialized",
    },
  ],
}
`;

exports[`runs a net with "and" split and "and" join 2`] = `
{
  "conditions": [
    {
      "marking": 1,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:scan_goods->pay",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pay->pack_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pay->issue_receipt",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pack_goods->check_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:issue_receipt->check_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
  ],
  "tasks": [
    {
      "generation": 0,
      "name": "scan_goods",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "pay",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "pack_goods",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "issue_receipt",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "check_goods",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "checkout",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`runs a net with "and" split and "and" join 3`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 1,
      "name": "implicit:scan_goods->pay",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pay->pack_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pay->issue_receipt",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pack_goods->check_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:issue_receipt->check_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "scan_goods",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "pay",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "pack_goods",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "issue_receipt",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "check_goods",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "checkout",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`runs a net with "and" split and "and" join 4`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:scan_goods->pay",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 1,
      "name": "implicit:pay->pack_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 1,
      "name": "implicit:pay->issue_receipt",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pack_goods->check_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:issue_receipt->check_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "scan_goods",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 1,
      "name": "pay",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "pack_goods",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "issue_receipt",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "check_goods",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "checkout",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`runs a net with "and" split and "and" join 5`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:scan_goods->pay",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pay->pack_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 1,
      "name": "implicit:pay->issue_receipt",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 1,
      "name": "implicit:pack_goods->check_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:issue_receipt->check_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "scan_goods",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 1,
      "name": "pay",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 1,
      "name": "pack_goods",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "issue_receipt",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "check_goods",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "checkout",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`runs a net with "and" split and "and" join 6`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:scan_goods->pay",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pay->pack_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pay->issue_receipt",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 1,
      "name": "implicit:pack_goods->check_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 1,
      "name": "implicit:issue_receipt->check_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "scan_goods",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 1,
      "name": "pay",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 1,
      "name": "pack_goods",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 1,
      "name": "issue_receipt",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 0,
      "name": "check_goods",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "checkout",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`runs a net with "and" split and "and" join 7`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 1,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:scan_goods->pay",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pay->pack_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pay->issue_receipt",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:pack_goods->check_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "marking": 0,
      "name": "implicit:issue_receipt->check_goods",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "scan_goods",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 1,
      "name": "pay",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 1,
      "name": "pack_goods",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 1,
      "name": "issue_receipt",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
    {
      "generation": 1,
      "name": "check_goods",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "checkout",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "checkout",
      "parent": null,
      "state": "completed",
    },
  ],
}
`;
